src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/24.diff.sql

```diff
@@ -0,0 +1,152 @@
+
+    WITH PartitionBoundaries
+    AS (
+        /* Normal logic when prior watermark exists, grab prior partition to ensure we do not miss data that was written across a partition boundary,
+           and includes partitions until current one to ensure we keep moving to the next partition when some partitions do not have any resource changes. */
+        SELECT CAST(prv.value as datetime2(7)) AS PartitionBoundary FROM sys.partition_range_values AS prv WITH (NOLOCK)
+            INNER JOIN sys.partition_functions AS pf WITH (NOLOCK) ON pf.function_id = prv.function_id
+        WHERE pf.name = N'PartitionFunction_ResourceChangeData_Timestamp'
+            AND SQL_VARIANT_PROPERTY(prv.Value, 'BaseType') = 'datetime2'
+            AND CAST(prv.value AS datetime2(7)) BETWEEN @precedingPartitionBoundary AND @endDateTimeToFilter
+    )
+    SELECT TOP(@pageSize) Id,
+        Timestamp,
+        ResourceId,
+        ResourceTypeId,
+        ResourceVersion,
+        ResourceChangeTypeId
+    FROM PartitionBoundaries AS p
+    CROSS APPLY (
+        /* Given the fact that Read Committed Snapshot isolation level is enabled on the FHIR database,
+           using TABLOCK and HOLDLOCK table hints to avoid skipping resource changes 
+           due to interleaved transactions on the resource change data table. */
+        SELECT TOP(@pageSize) Id,
+            Timestamp,
+            ResourceId,
+            ResourceTypeId,
+            ResourceVersion,
+            ResourceChangeTypeId
+        /* Acquires and holds the table lock to prevent new resource changes from being created during the select query execution.
+           Without the TABLOCK and HOLDLOCK hints, the lock will be applied at the row or page level which can result in missed reads
+           if an insert occurs in range that was previously read and the locks were released on the read. */
+        FROM dbo.ResourceChangeData WITH (TABLOCK, HOLDLOCK)
+            WHERE Id >= @startId
+                AND $PARTITION.PartitionFunction_ResourceChangeData_Timestamp(Timestamp) = $PARTITION.PartitionFunction_ResourceChangeData_Timestamp(p.PartitionBoundary)
+        ORDER BY Id ASC
+        ) AS rcd
+    ORDER BY rcd.Id ASC;
+END;
+GO
+
+IF EXISTS (SELECT 1 FROM sys.indexes WHERE name = 'PK_ResourceChangeData_TimestampId')
+BEGIN
+    EXEC dbo.LogSchemaMigrationProgress 'Deleting PK_ResourceChangeData_TimestampId index from ResourceChangeData table.';
+
+    /* Drops PK_ResourceChangeData_TimestampId index. "ONLINE = ON" indicates long-term table locks aren't held for the duration of the index operation. 
+       During the main phase of the index operation, only an Intent Share (IS) lock is held on the source table. 
+       This behavior enables queries or updates to the underlying table and indexes to continue. */
+    ALTER TABLE dbo.ResourceChangeData DROP CONSTRAINT PK_ResourceChangeData_TimestampId WITH(ONLINE = ON);
```

Comment:
Is the primary key which we are dropping here or below clustered or nonclustered. Since if it is nonclustered, would it make sense to create clustered index first and then drop primary key for both the tables.

Reply:
The existing primary key clustered is clustered primary key index. So, we have to drop first before creating another clustered index.

Reply:
Thanks for confirming it!